src/provers/sharp/mod.rs

```diff
@@ -0,0 +1,96 @@
+use crate::settings::SettingsProvider;
+
+pub const SHARP_SETTINGS_NAME: &str = "sharp";
+
+/// SHARP (aka GPS) is a shared proving service hosted by Starkware.
+pub struct SharpProverService {
+    sharp_client: SharpClient,
+    fact_checker: FactChecker,
+    task_facts: HashMap<TaskId, B256>,
+}
+
+#[async_trait]
+impl ProverService for SharpProverService {
+    async fn submit_task(&self, task: Task) -> Result<TaskId, ProverServiceError> {
+        let encoded_pie = match task {
+            Task::EncodedPie(enc) => enc,
+            Task::PieObject(pie) => {
+                todo!("Need to resolve cairo-vm versions comflict");
+                // pie::encode_pie_mem(*pie).map_err(SharpError::PieEncode)?
+            }
+        };
+        let res = self.sharp_client.add_job(&encoded_pie).await?;
+        if let Some(job_key) = res.cairo_job_key {
+            Ok(job_key.to_string())
+        } else {
+            Err(ProverServiceError::TaskInvalid(res.error_message.unwrap_or_default()).into())
+        }
+    }
+
+    async fn get_task_status(&self, task_id: &TaskId) -> Result<TaskStatus, ProverServiceError> {
+        let job_key = Uuid::from_str(task_id).map_err(SharpError::TaskIdParse)?;
+        let res = self.sharp_client.get_job_status(&job_key).await?;
+        match res.status {
+            CairoJobStatus::FAILED => Ok(TaskStatus::Failed(res.error_log.unwrap_or_default())),
+            CairoJobStatus::INVALID => {
+                Ok(TaskStatus::Failed(format!("Task is invalid: {:?}", res.invalid_reason.unwrap_or_default())))
+            }
+            CairoJobStatus::UNKNOWN => Ok(TaskStatus::Failed(format!("Task not found: {}", task_id))),
+            CairoJobStatus::IN_PROGRESS | CairoJobStatus::NOT_CREATED | CairoJobStatus::PROCESSED => {
+                Ok(TaskStatus::Processing)
+            }
+            CairoJobStatus::ONCHAIN => {
+                let fact = self.get_fact(task_id).ok_or(ProverServiceError::FactNotFound(task_id.clone()))?;
+                if self.fact_checker.is_valid(fact).await? {
+                    Ok(TaskStatus::Succeeded)
+                } else {
+                    Ok(TaskStatus::Failed(format!("Fact {} is not valid or not registed", hex::encode(&fact))))
+                }
+            }
+        }
+    }
+}
+
+impl SharpProverService {
+    pub fn new(sharp_client: SharpClient, fact_checker: FactChecker) -> Self {
+        Self { sharp_client, fact_checker, task_facts: HashMap::new() }
+    }
+
+    pub fn with_settings(settings: &impl SettingsProvider) -> Self {
+        let sharp_cfg: SharpConfig = settings.get_settings(SHARP_SETTINGS_NAME).unwrap();
+        let sharp_client = SharpClient::new(sharp_cfg.service_url);
+        let fact_checker = FactChecker::new(sharp_cfg.rpc_node_url, sharp_cfg.verifier_address);
+        Self::new(sharp_client, fact_checker)
+    }
+
+    pub fn set_fact(&mut self, task_id: &TaskId, cairo_pie: &CairoPie) -> Result<(), SharpError> {
```

Comment:
`set_fact` isn't being called right now. should it be called inside `submit_task`?